File: HeraldTest-Flutter2.0/ios/Runner/DistanceEstimator.swift
import Herald
import Darwin

// Note: "model" is used both for SmoothedLinerModel and ModelWrapper.

public class DistanceEstimator {
    private var modelMap: [Int: ModelWrapper] = [:]
    private var tag: String = "DistanceEstimator"

    public init() {

    }

    // Removes the model associated with the given UUID
    public func removeModel(_ UUID: Int) {
        print("\(tag) Removed UUID: \(UUID)")
        self.modelMap.removeValue(forKey: UUID)
    }

    // Creates a model for a given UUID
    private func createModel(_ UUID: Int, _ phone: Int) -> ModelWrapper {
        let model = ModelWrapper(phoneCode: phone)
        self.modelMap[UUID] = model
        return model
    }

    // Adds a RSSI value to the model handling the given UUID (will create a new model if needed)
    public func addRSSI(_ UUID: Int, _ rssi: Double, _ phone: Int) {
        let sample = Sample(value: RSSI(rssi))
        let model = self.modelMap[UUID] ?? self.createModel(UUID, phone)
        model.addSample(sample)
    }

    public func getDistance(_ UUID: Int) -> Double? { // TODO deal with not having enough samples to give distance or no model exists
        if let model = self.modelMap[UUID] {
            return model.getDistance()
        } else {
            return nil
        }
    }

    public func getSampleCount(_ UUID: Int) -> Int {
        if let model = self.modelMap[UUID] {
            return model.getSampleCount()
        } else {
            return 0
        }
    }
    
    public func getMedianRSSI(_ UUID: Int) -> Double? {
        if let model = self.modelMap[UUID] {
            return model.getMedianRSSI()
        } else {
            return nil
        }
    }

    // This class essentially takes the role of SmoothedLinearModelAnalyzer and is heavily influenced by it
    // I made this class since SmoothedLinearModelAnalyzer is not *quite* what is required for this use
    private class ModelWrapper {

        var lastUpdated: Date
        var lastDistance: Double
        var model: CoarseDistanceModel
//        var filter: SimpleKalmanFilter

        var window: SampleList
        var minimumWindowSize: Int
        var maximumWindowSize: Int // TODO decide where these values go and how they work
        var smoothingWindow: TimeInterval


        init(phoneCode: Int) {
            self.lastUpdated = Date()
            self.minimumWindowSize = 10
            self.maximumWindowSize = 50
            self.smoothingWindow = TimeInterval(60)
            self.window = SampleList(maximumWindowSize)
            self.model = CoarseDistanceModel(phoneCode: phoneCode)
//            self.filter = SimpleKalmanFilter(1, 1, 0.03)
            
            self.lastDistance = 0
        }

        func addSample(_ sample: Sample) {
            window.push(sample: sample)
        }

        func getSampleCount() -> Int {
            return window.size()
        }
        
        func getMedianRSSI() -> Double? {
            return self.model.medianOfRssi()
        }

        func getDistance() -> Double? { // TODO fix breaking for very close ranges
            let timeNow = Date()
            // Remove stale samples
            self.window.clearBeforeDate(Date(timeInterval: -self.smoothingWindow, since: timeNow))

            if self.window.size() < minimumWindowSize {
               return nil
            }
            
            self.model.reset()
            for i in 0...self.window.size()-1 {
                if let sample = self.window.get(i) {
                    self.model.map(value: sample)
                }
            }
            
            let distanceBucket = self.model.reduce()
//            let distanceEstimation =  self.filter.updateEstimate(distanceMeasurement)
            return distanceBucket
        }
    }

}

File: HeraldTest-Flutter2.0/android/app/src/main/java/com/example/herald_flutter/DistanceEstimator.java
package com.example.herald_flutter;

import androidx.annotation.NonNull;

import io.heraldprox.herald.sensor.datatype.Date;
import io.heraldprox.herald.sensor.datatype.RSSI;
import io.heraldprox.herald.sensor.analysis.sampling.Sample;
import io.heraldprox.herald.sensor.analysis.sampling.SampleList;
import io.heraldprox.herald.sensor.datatype.TimeInterval;
import io.flutter.Log;

import java.util.Hashtable;
import java.util.Map;

// Note: "model" is used both for PiecewiseDistanceModel and ModelWrapper.

public class DistanceEstimator { 
    private Map<Integer, ModelWrapper> modelMap;
    private static final String tag = "DistanceEstimator";

    public DistanceEstimator() {
        modelMap = new Hashtable<Integer, ModelWrapper>();
    }

    // Removes the model associated with the given UUID
    public void removeModel(int UUID) {
        Log.i(tag, "Removed UUID: " + UUID);
        modelMap.remove(UUID);
    }
    
    // Creates a model for a given UUID
    private ModelWrapper createModel(int UUID, int phone) {
        ModelWrapper model = new ModelWrapper(phone);
        modelMap.put(UUID, model);
        return model;
    }

    // Adds a RSSI value to the model handling the given UUID (will create a new model if needed)
    public void addRSSI(int UUID, Double rssi, int phone) {
        Sample<RSSI> sample = new Sample(new RSSI(rssi));
        ModelWrapper model = modelMap.get(UUID);
        if (model == null) {
            model = createModel(UUID, phone);
        }
        model.addSample(sample);
    }

    public Double getDistance(int UUID) {
        ModelWrapper model = modelMap.get(UUID);
        if (model != null) {
            return model.getDistance();
        } else {
            return null;
        }        
    }

    public int getSampleCount(int UUID) {
        ModelWrapper model = modelMap.get(UUID);
        if (model != null) {
            return model.getSampleCount();
        } else {
            return 0;
        }
    }

    public Double getMedianRSSI(int UUID) {
        ModelWrapper model = modelMap.get(UUID);
        if (model != null) {
            return model.getMedianRSSI();
        } else {
            return null;
        }          
    }

    public Double getKalmandRSSI(int UUID) {
        ModelWrapper model = modelMap.get(UUID);
        if (model != null) {
            return model.getKalmanRSSI();
        } else {
            return null;
        }          
    }

    private class ModelWrapper {

        Date lastUpdated;
        Double lastDistance;
        private CoarseDistanceModel model;
        private SampleList<RSSI> window;
        int minimumWindowSize;
        int maximumWindowSize;
        TimeInterval smoothingWindow;

        public ModelWrapper(int phoneCode) {
            lastUpdated = new Date();
            minimumWindowSize = 10;
            maximumWindowSize = 50;
            smoothingWindow = new TimeInterval(60);
            window = new SampleList<RSSI>(maximumWindowSize);
            model = new CoarseDistanceModel(phoneCode);
        }

        public void addSample(Sample<RSSI> sample) {
            window.push(sample);
        }

        public int getSampleCount() {
            return window.size();
        }

        public Double getMedianRSSI() {
            return model.medianOfRssi();
        }

        public Double getKalmanRSSI() {
            return model.kalmanOfRssi();
        }

        public Double getDistance() {
            Date timeNow = new Date();
            // Remove stale samples
            window.clearBeforeDate(new Date(timeNow.secondsSinceUnixEpoch() - smoothingWindow.value));

            if (window.size() < minimumWindowSize) {
                // Not enough samples
                return null;
            }

            model.reset();
            for (Sample<RSSI> sample : window) {
                model.map(sample);
            }

            Double distanceBucket = model.reduce();
            return distanceBucket;
        }
    }
}
